### PR TITLE
Map user memory passed to accept() in kernel build

### PR DESCRIPTION
**Pull request by gpoulios**
## Summary

At least on RISC-V, TCP server sockets were pretty much unusable on high load environments and kernel build.

User addresses passed to `accept()` would be accessed when the wrong MMU mappings were active. A crash would manifest when attempting to `accept()` for instance on a TCP server socket under significant load. The accept event handler would be called by the HP worker upon client connection. At this point, `accept_tcpsender()` would attempt to write to `addr` resulting in a page fault. Reproducibility would depend on the current system load (num tasks or CPU stress) but in loaded environments, it would crash almost 100% of the times.

## Impact

The change should be transparent to users. `accept()` now maps pointer arguments to kernel virtual memory before proceeding. It also checks whether the given pointers indeed point to user memory.

It should be noted that Linux does this the other way around: it operates on kernel stack allocated data and once done, it copies them to user. This can also be a viable alternative, albeit with one extra copy and a little extra memory.

### RISC-V testing only

Prior to patch: trying with `CONFIG_NETUTILS_NETCAT=y` would cause a store fault once a client attempted to connect. The point of the crash was consistently `tcp_accept.c / accept_tcpsender()` at the first write to the `inaddr` structure ([line 102](https://github.com/apache/nuttx/blob/5d68ab635c065d97ac0cbc08372be7c37fb59d02/net/tcp/tcp_accept.c#L102)).

After the patch: `accept()` works normally after multiple connections.

### CODE REVIEW

File: fs/socket/accept.c [hunk 1]
```diff
@@ -38,10 +38,62 @@
...
+
+      else if (*p_kaddr == uaddr)
+        {
+          return -EFAULT;
+        }
+    }
+
+  return OK;
+}
+#endif
```
Comment (pussuw) [inline fs/socket/accept.c:95]: This wrapper might be unnecessary, is there some kind of POSIX compliance requirement for returning -EFAULT ? I think the return value could always be -ENOMEM (meaning no free space to map the user vaddr).

**Review by pussuw — COMMENTED** [review]
One question about the wrapper func, otherwise this looks good. Please upstream it first so we don't have to sort out any conflicts if the patch requires changes there!

File: fs/socket/accept.c (hunk 1 above)
Comment (gpoulios) [reply fs/socket/accept.c:95]: I got the idea for the wrapper from [this](https://github.com/tiiuae/nuttx/blob/bd5afce6ff35be04086e5839bbd75ad48d6bb47b/sched/semaphore/sem_holder.c#L101).

Regarding `EFAULT`, I believe it's not in the posix standard, only in the linux implementation, and also aligned with `accept4`'s documentation a few lines below (see line 159). The error is not about free space, but rather the user trying to abuse the syscall to write to memory they don't have access to.

**Comment by gpoulios** [comment]
cherry picked from upstream https://github.com/apache/nuttx/pull/15278

**Pull request merged**